feat(csharp,debug): add C# debugger support

filipdutescu opened the pull request:
Add C# debugger support in the `languages.toml` file. This includes "launch" and "attach" configurations that employ [netcoredbg][netdbg].

In order to make it work with the aforementioned debugger, the `Initialized` DAP event response is changed to include an optional `DebuggerCapabilities` response, as per the specification.

[netdbg]: https://github.com/Samsung/netcoredbg

Closes: #4212
Signed-off-by: Filip Dutescu <filip.dutescu@gmail.com>

the-mikedavis commented on languages.toml:
```diff
@@ -233,6 +233,25 @@ comment-token = "//"
+name = "launch"
+request = "launch"
+completion = [ { name = "path to dll", completion = "filename" } ]
+args = { type="coreclr", console = "internalConsole", internalConsoleOptions = "openOnSessionStart", program = "{0}" }
```
```suggestion
args = { type = "coreclr", console = "internalConsole", internalConsoleOptions = "openOnSessionStart", program = "{0}" }
```

filipdutescu commented:
I do not know what the proper procedure for addressing comments is, but I pushed one specifically with your change and if approved, I can rebase and squash my PR prior to merger, if it help or you are ok with this flow.

filipdutescu commented:
What does `helix.log` say? I ran debug without problem on a small CLI C# project and all worked well :thinking: where did you set the breakpoint? That could also play a part in your issue.

CptPotato commented:
Thanks for the quick reply! This is the test program:

```cs
using System;

namespace Test;

static class Program
{
    public static void Main()
    {
        var x = "asdf";
        Console.WriteLine(x); // < breakpoint on this line
    }
}
```

You're right, the helix log reports an error for an unhandled event. Unless I messed something up my build should include the code changes from this PR. I'll double check again just to be sure.

<details>
<summary>helix.log</summary>

```js
2022-10-12T11:41:35.057 helix_view::handlers::dap [WARN] Unhandled event Capabilities(Capabilities {
    capabilities: DebuggerCapabilities {
        supports_configuration_done_request: Some(true),
        supports_function_breakpoints: Some(true),
        supports_conditional_breakpoints: Some(true),
        supports_hit_conditional_breakpoints: None,
        supports_evaluate_for_hovers: None,
        supports_step_back: None,
        supports_set_variable: Some(true),
        supports_restart_frame: None,
        supports_goto_targets_request: None,
        supports_step_in_targets_request: None,
        supports_completions_request: None,
        supports_modules_request: None,
        supports_restart_request: None,
        supports_exception_options: Some(false),
        supports_value_formatting_options: None,
        supports_exception_info_request: Some(true),
        support_terminate_debuggee: Some(true),
        support_suspend_debuggee: None,
        supports_delayed_stack_trace_loading: None,
        supports_loaded_sources_request: None,
        supports_log_points: None,
        supports_terminate_threads_request: None,
        supports_set_expression: Some(true),
        supports_terminate_request: Some(true),
        supports_data_breakpoints: None,
        supports_read_memory_request: None,
        supports_write_memory_request: None,
        supports_disassemble_request: None,
        supports_cancel_request: Some(true),
        supports_breakpoint_locations_request: None,
        supports_clipboard_context: None,
        supports_stepping_granularity: None,
        supports_instruction_breakpoints: None,
        supports_exception_filter_options: Some(true),
        exception_breakpoint_filters: Some([ExceptionBreakpointsFilter {
            filter: "all",
            label: "all",
            description: None,
            default: None,
            supports_condition: None,
            condition_description: None
        }, ExceptionBreakpointsFilter {
            filter: "user-unhandled",
            label: "user-unhandled",
            description: None,
            default: None,
            supports_condition: None,
            condition_description: None
        }]),
        completion_trigger_characters: None,
        additional_module_columns: None,
        supported_checksum_algorithms: None
    }
})
2022-10-12T11:41:35.093 helix_dap::transport [ERROR] err: <- IO(Os {
    code: 10054,
    kind: ConnectionReset,
    message: "An existing connection was forcibly closed by the remote host."
})
```
</details>

filipdutescu commented:
Try running `helix` with the `-v` param to get more verbose logs. Also, I forgot to ask but could you tell me what your environment looks like (OS, `netcoredbg` version etc)? I have no problems on ArchLinux, netcoredbg 2.0.0-78.

filipdutescu commented:
Here is also my test program which worked:

```csharp
namespace TestHx
{
	public class Program
	{
		public static void Main(string[] args)
		{
			Console.WriteLine("Enter text:");
			var a = 123;
			// Thread.Sleep(10000);
			// var a = Console.ReadLine();
			Console.WriteLine("Hello, World!"); // breakpoint here
			Console.WriteLine($"{a}");
		}
	}
}
```

Make sure you also built the project right before launching the debugger, in order to have the latest version available, otherwise breakpoints could be skipped since the source code and DLL fail to match

CptPotato commented:
So I did a cargo clean & rebuild but get the same behavior.

Also, here's my env:
- Windows 10 x64 (maybe that's the problem 🙃)
- helix built from `master` (with changes from this PR applied)
- netcoredbg `2.2.0-7 (1c7c1f5, Release)`

This is the verbose log (only including `helix_dap` and `helix_view::handlers::dap`).
The suspicious part is that the DAP connection is closed (towards the end of the log) so it might have exited early or crashed.

<details>
<summary>helix.log</summary>

```js
helix_dap::transport[INFO] -> DAP {
    "type": "request",
    "seq": 0,
    "command": "initialize",
    "arguments": {
        "adapterID": "netcoredbg",
        "clientID": "hx",
        "clientName": "helix",
        "columnsStartAt1": true,
        "linesStartAt1": true,
        "locale": "en-us",
        "pathFormat": "path",
        "supportsInvalidatedEvent": false,
        "supportsMemoryReferences": false,
        "supportsProgressReporting": false,
        "supportsRunInTerminalRequest": true,
        "supportsVariablePaging": false,
        "supportsVariableType": true
    }
}
helix_dap::transport[INFO] <- DAP {
    "body": {
        "capabilities": {
            "exceptionBreakpointFilters": [{
                "filter": "all",
                "label": "all"
            }, {
                "filter": "user-unhandled",
                "label": "user-unhandled"
            }],
            "supportTerminateDebuggee": true,
            "supportsCancelRequest": true,
            "supportsConditionalBreakpoints": true,
            "supportsConfigurationDoneRequest": true,
            "supportsExceptionFilterOptions": true,
            "supportsExceptionInfoRequest": true,
            "supportsExceptionOptions": false,
            "supportsFunctionBreakpoints": true,
            "supportsSetExpression": true,
            "supportsSetVariable": true,
            "supportsTerminateRequest": true
        }
    },
    "event": "capabilities",
    "seq": "1",
    "type": "event"
}
helix_dap::transport[INFO] <- DAP event Capabilities(Capabilities {
    capabilities: DebuggerCapabilities {
        supports_configuration_done_request: Some(true),
        supports_function_breakpoints: Some(true),
        supports_conditional_breakpoints: Some(true),
        supports_hit_conditional_breakpoints: None,
        supports_evaluate_for_hovers: None,
        supports_step_back: None,
        supports_set_variable: Some(true),
        supports_restart_frame: None,
        supports_goto_targets_request: None,
        supports_step_in_targets_request: None,
        supports_completions_request: None,
        supports_modules_request: None,
        supports_restart_request: None,
        supports_exception_options: Some(false),
        supports_value_formatting_options: None,
        supports_exception_info_request: Some(true),
        support_terminate_debuggee: Some(true),
        support_suspend_debuggee: None,
        supports_delayed_stack_trace_loading: None,
        supports_loaded_sources_request: None,
        supports_log_points: None,
        supports_terminate_threads_request: None,
        supports_set_expression: Some(true),
        supports_terminate_request: Some(true),
        supports_data_breakpoints: None,
        supports_read_memory_request: None,
        supports_write_memory_request: None,
        supports_disassemble_request: None,
        supports_cancel_request: Some(true),
        supports_breakpoint_locations_request: None,
        supports_clipboard_context: None,
        supports_stepping_granularity: None,
        supports_instruction_breakpoints: None,
        supports_exception_filter_options: Some(true),
        exception_breakpoint_filters: Some([ExceptionBreakpointsFilter {
            filter: "all",
            label: "all",
            description: None,
            default: None,
            supports_condition: None,
            condition_description: None
        }, ExceptionBreakpointsFilter {
            filter: "user-unhandled",
            label: "user-unhandled",
            description: None,
            default: None,
            supports_condition: None,
            condition_description: None
        }]),
        completion_trigger_characters: None,
        additional_module_columns: None,
        supported_checksum_algorithms: None
    }
})
helix_dap::transport[INFO] <- DAP {
    "body": {},
    "event": "initialized",
    "seq": "2",
    "type": "event"
}
helix_dap::transport[INFO] <- DAP event Initialized(Some(DebuggerCapabilities {
    supports_configuration_done_request: None,
    supports_function_breakpoints: None,
    supports_conditional_breakpoints: None,
    supports_hit_conditional_breakpoints: None,
    supports_evaluate_for_hovers: None,
    supports_step_back: None,
    supports_set_variable: None,
    supports_restart_frame: None,
    supports_goto_targets_request: None,
    supports_step_in_targets_request: None,
    supports_completions_request: None,
    supports_modules_request: None,
    supports_restart_request: None,
    supports_exception_options: None,
    supports_value_formatting_options: None,
    supports_exception_info_request: None,
    support_terminate_debuggee: None,
    support_suspend_debuggee: None,
    supports_delayed_stack_trace_loading: None,
    supports_loaded_sources_request: None,
    supports_log_points: None,
    supports_terminate_threads_request: None,
    supports_set_expression: None,
    supports_terminate_request: None,
    supports_data_breakpoints: None,
    supports_read_memory_request: None,
    supports_write_memory_request: None,
    supports_disassemble_request: None,
    supports_cancel_request: None,
    supports_breakpoint_locations_request: None,
    supports_clipboard_context: None,
    supports_stepping_granularity: None,
    supports_instruction_breakpoints: None,
    supports_exception_filter_options: None,
    exception_breakpoint_filters: None,
    completion_trigger_characters: None,
    additional_module_columns: None,
    supported_checksum_algorithms: None
}))
helix_dap::transport[INFO] <- DAP {
    "body": {
        "exceptionBreakpointFilters": [{
            "filter": "all",
            "label": "all"
        }, {
            "filter": "user-unhandled",
            "label": "user-unhandled"
        }],
        "supportTerminateDebuggee": true,
        "supportsCancelRequest": true,
        "supportsConditionalBreakpoints": true,
        "supportsConfigurationDoneRequest": true,
        "supportsExceptionFilterOptions": true,
        "supportsExceptionInfoRequest": true,
        "supportsExceptionOptions": false,
        "supportsFunctionBreakpoints": true,
        "supportsSetExpression": true,
        "supportsSetVariable": true,
        "supportsTerminateRequest": true
    },
    "command": "initialize",
    "request_seq": 0,
    "seq": "3",
    "success": true,
    "type": "response"
}
helix_dap::transport[INFO] <- DAP success in response to 0
helix_view::handlers::dap[WARN] Unhandled event Capabilities(Capabilities {
    capabilities: DebuggerCapabilities {
        supports_configuration_done_request: Some(true),
        supports_function_breakpoints: Some(true),
        supports_conditional_breakpoints: Some(true),
        supports_hit_conditional_breakpoints: None,
        supports_evaluate_for_hovers: None,
        supports_step_back: None,
        supports_set_variable: Some(true),
        supports_restart_frame: None,
        supports_goto_targets_request: None,
        supports_step_in_targets_request: None,
        supports_completions_request: None,
        supports_modules_request: None,
        supports_restart_request: None,
        supports_exception_options: Some(false),
        supports_value_formatting_options: None,
        supports_exception_info_request: Some(true),
        support_terminate_debuggee: Some(true),
        support_suspend_debuggee: None,
        supports_delayed_stack_trace_loading: None,
        supports_loaded_sources_request: None,
        supports_log_points: None,
        supports_terminate_threads_request: None,
        supports_set_expression: Some(true),
        supports_terminate_request: Some(true),
        supports_data_breakpoints: None,
        supports_read_memory_request: None,
        supports_write_memory_request: None,
        supports_disassemble_request: None,
        supports_cancel_request: Some(true),
        supports_breakpoint_locations_request: None,
        supports_clipboard_context: None,
        supports_stepping_granularity: None,
        supports_instruction_breakpoints: None,
        supports_exception_filter_options: Some(true),
        exception_breakpoint_filters: Some([ExceptionBreakpointsFilter {
            filter: "all",
            label: "all",
            description: None,
            default: None,
            supports_condition: None,
            condition_description: None
        }, ExceptionBreakpointsFilter {
            filter: "user-unhandled",
            label: "user-unhandled",
            description: None,
            default: None,
            supports_condition: None,
            condition_description: None
        }]),
        completion_trigger_characters: None,
        additional_module_columns: None,
        supported_checksum_algorithms: None
    }
})
helix_dap::transport[INFO] -> DAP {
    "type": "request",
    "seq": 2,
    "command": "setBreakpoints",
    "arguments": {
        "breakpoints": [{
            "line": 10
        }],
        "source": {
            "path": "[...]\\test\\Program.cs"
        },
        "sourceModified": false
    }
}
helix_dap::transport[ERROR] err: <- IO(Os {
    code: 10054,
    kind: ConnectionReset,
    message: "An existing connection was forcibly closed by the remote host."
})
helix_dap::transport[INFO] -> DAP {
    "type": "request",
    "seq": 3,
    "command": "configurationDone",
    "arguments": null
}
```
</details>

---

Using the `netcoredbg` command line directly seems to work, though:

<details>
<summary>helix.log</summary>

```
ncdb> file dotnet
ncdb> set args test.dll
ncdb> break test.dll!Program.cs:10
 Breakpoint 1 at Program.cs:10 --pending, warning: No executable code of the debugger's target code type is associated with this line.
ncdb> run
^running

library loaded: [...]\dotnet\shared\Microsoft.NETCore.App\6.0.9\System.Private.CoreLib.dll
no symbols loaded, base address: 0x7ffc616b0000, size: 10629120(0xa23000)

thread created, id: 16156

library loaded: [...]\test\bin\Debug\net6.0\test.dll
symbols loaded, base address: 0x1a9dc7c0000, size: 32768(0x8000)
breakpoint modified,  Breakpoint 1 at [..]\test\Program.cs:10

library loaded: [...]\dotnet\shared\Microsoft.NETCore.App\6.0.9\System.Runtime.dll
no symbols loaded, base address: 0x1a9dc7d0000, size: 57344(0xe000)

library loaded: [...]\dotnet\shared\Microsoft.NETCore.App\6.0.9\System.Console.dll
no symbols loaded, base address: 0x7ffca3480000, size: 155648(0x26000)

stopped, reason: breakpoint 1 hit, thread id: 16156, stopped threads: all, times= 1, frame={Test.Program.Main() at [..]\test\Program.cs:10
}
ncdb> print x
x = "asdf"
ncdb> detach

stopped, reason: exited, exit-code: 0
^exit
```
</details>

filipdutescu commented:
At a first glance, it seems to be a `netcoredbg` issue, rather than a Helix one :thinking: I don't have access to a Windows machine right now, but later on today I could try and setup one. It also seems to have been caused by `netcoredbg` not liking the `SetBreakpoints` request and crashed.

Are you able to debug Helix on your environment? I'd be really curious if we could get any more info from it, besides what's logged. Also, by any chance, would you happen to have a Linux/MacOS machine on which you could test it as well?

filipdutescu commented:
I think the path is wrong:

```rust
helix_dap::transport[INFO] -> DAP {
    "type": "request",
    "seq": 2,
    "command": "setBreakpoints",
    "arguments": {
        "breakpoints": [{
            "line": 10
        }],
        "source": {
            "path": "[...]\\test\\Program.cs"
        },
        "sourceModified": false
    }
}
```

For me it actually puts the absolute path there. @the-mikedavis any idea why that might happen on Win?

CptPotato commented:
Thanks a lot for looking into this.

> Are you able to debug Helix on your environment? I'd be really curious if we could get any more info from it, besides what's logged. Also, by any chance, would you happen to have a Linux/MacOS machine on which you could test it as well?

I didn't look much into setting up rust debugging yet to be honest. I could try to get it working or otherwise add a few additional logs to narrow it down.
I also have a Linux machine but I never compiled C# on it, though I might give it a shot if I can set it up. Would be interesting to see if the problem is platform specific.

In the meantime, I don't mean to block this PR in case it works on Linux/MacOS. There's probably people other than me that have been looking forward to C# debugging in helix.

---

> I think the path is wrong

Ah my bad.. I didn't mention that I truncated all file paths in the logs I have posted. It is indeed the full path to the source file.

filipdutescu commented:
How did you get `netcoredbg`? Could you try using the instructions from https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=netcoredbg (Win equivalent to the `build()` section)?

the-mikedavis commented on helix-dap/src/types.rs:
```diff
@@ -726,7 +726,7 @@ pub mod events {
     // seq is omitted as unused and is not sent by some implementations
     pub enum Event {
-        Initialized,
+        Initialized(Option<DebuggerCapabilities>),
```
👍 Looks like Initialized extends Event which has an optional body field https://microsoft.github.io/debug-adapter-protocol/specification#Base_Protocol_Event

the-mikedavis reviewed (approved):
This works for me locally on linux 👍
